core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/GroupByAggregator.java

```diff
@@ -107,24 +108,31 @@ public class GroupByAggregator extends RecordReducibleAggregator<AggregateResult
     put("ANYIF", "AnyIf");
   }};
 
+  // Ansi SQL aggregations
   private final HashMap<GroupByConfig.Function, String> functionSqlMap =
     new HashMap<GroupByConfig.Function, String>() {{
-    put(GroupByConfig.Function.AVG, "AVG(%s)");
-    put(GroupByConfig.Function.COUNT, "COUNT(%s)");
-    put(GroupByConfig.Function.MAX, "MAX(%s)");
-    put(GroupByConfig.Function.MIN, "MIN(%s)");
-    put(GroupByConfig.Function.STDDEV, "STDDEV_POP(%s)");
-    put(GroupByConfig.Function.SUM, "SUM(%s)");
-    put(GroupByConfig.Function.VARIANCE, "VAR_POP(%s)");
-    put(GroupByConfig.Function.COLLECTLIST, "ARRAY_AGG(%s IGNORE NULLS)");
-    put(GroupByConfig.Function.COLLECTSET, "ARRAY_AGG(DISTINCT %s IGNORE NULLS)");
-    put(GroupByConfig.Function.COUNTDISTINCT, "COUNT(DISTINCT %s) + IF(COUNTIF(%<s IS NULL) > 0, 1, 0)");
-    put(GroupByConfig.Function.COUNTNULLS, "COUNTIF(%s IS NULL)");
-    put(GroupByConfig.Function.CONCAT, "STRING_AGG(CAST(%s AS STRING), \", \")");
-    put(GroupByConfig.Function.CONCATDISTINCT, "STRING_AGG(DISTINCT CAST(%s AS STRING) , \", \")");
-    put(GroupByConfig.Function.LOGICALAND, "COALESCE(LOGICAL_AND(%s), TRUE)");
-    put(GroupByConfig.Function.LOGICALOR, "COALESCE(LOGICAL_OR(%s), FALSE)");
-  }};
+      put(GroupByConfig.Function.AVG, "AVG(%s)");
+      put(GroupByConfig.Function.MAX, "MAX(%s)");
+      put(GroupByConfig.Function.MIN, "MIN(%s)");
+      put(GroupByConfig.Function.STDDEV, "STDDEV_POP(%s)");
+      put(GroupByConfig.Function.SUM, "SUM(%s)");
+      put(GroupByConfig.Function.VARIANCE, "VAR_POP(%s)");
+      put(GroupByConfig.Function.COUNT, "COUNT(%s)");
+      put(GroupByConfig.Function.COUNTNULLS, "SUM(CASE WHEN %s IS NULL THEN 1 ELSE 0 END)");
+      put(GroupByConfig.Function.COUNTDISTINCT,
+          "COUNT(DISTINCT %s) + CASE WHEN SUM(CASE WHEN %>s IS NULL THEN 1 ELSE 0 END) > 0 THEN 1 ELSE 0 END");
```

Comment:
nit: can this be `COUNT(DISTINCT %s) + MAX(SUM(CASE WHEN %>s IS NULL THEN 1 ELSE 0 END))` or it would produce a wrong result for empty set? Two nested `CASE` are hard to read

Reply:
This expression can be made safe by adding a Coalesce:

COUNT(DISTINCT %s) + COALESCE(MAX(CASE WHEN %>s IS NULL THEN 1 ELSE 0 END), 0)